text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,393 @@
+
+2. In certain cases when new, incompatible functionality is added to OTLP, the implementation of such new version needs to be able to detect that its peer is an implementation of an older version. In such case the newer implementation MUST perform the detection of peer's capability and adjust its behavior to match the expectation of older implementation. The detection is typically done by the presence or lack of certain field in the Protocol Buffer-encoded messages. The detection can also be based on the fact that new OTLP versions define new gRPC services and thus mere fact of calling certain services is enough to detect the version of the client.
+
+The current version of OTLP is the first version and it SHOULD NOT attempt to detect the capabilities of its peer and should operate as defined in this document.
+
+## Prior Art, Alternatives and Future Possibilities
+
+OTLP is an evolution of OpenCensus protocol based on the research and testing of its modifications in production at Omnition. The modifications include the addition of delivery acknowledgements and changes that make the protocol load-balancer friendly.
+
+OTLP uses Protocol Buffers for data encoding. Two other encodings were considered as alternative approaches: FlatBuffers and Capnproto. Both alternatives were rejected. FlatBuffers was rejected because it lacks required functionality in all languages except C++, particularly lack of verification of decoded data and inability to mutate in-memory data. Capnproto was rejected because it is not yet considered production ready, the API is not yet stable and like FlatBuffers it lacks ability mutate in-memory data.
+
+Both FlatBuffers and Capnproto are worth to be re-evaluated for future versions of OpenTelemetry protocol if they overcome currently known limitations.
+
+It is also worth researching transports other than gRPC. Other transports are not included in this RFC due to time limitations.
+
+Experimental implementation of OTLP over WebSockets exists and was researched as an alternate. This alternate will be considered to be included in a future OTLP Extensions RFC.
+
+## Open Questions
+
+One of the goals for telemetry protocol is reducing CPU usage and memory pressure in garbage collected languages. These goals were not address as part of this RFC and remain open. One of the promising future ways to address this is finding a more CPU and memory efficient encoding mechanism.
+
+Another goal for telemetry protocol is achieving high compression ratios for telemetry data while keeping CPU consumption low. This goal was not addressed as part of this RFC. 
+
+## Appendix A - Protocol Buffer Definitions
+
+This is Protocol Buffers schema for `Export` request and response:
+
+```
+// A request from client to server containing telemetry data to export.
+message ExportRequest {
+  // Unique sequential ID generated by the client.
+  uint64 id = 1;
+
+  // Telemetry data.
+  repeated Span spans = 2;
```

Comment:
This needs to be updated to reference a SpanBatch message that includes spans and node info.

Reply:
Possible alternate:
```
// A request from client to server containing telemetry data to export.
message ExportRequest {
  // Unique sequential ID generated by the client.
  uint64 id = 1;

  // Telemetry data.
  repeated NodeSpans nodeSpans = 2;
  repeated Metric metrics = 3;
}
message NodeSpans {
  Node node = 1;
  Resource resource = 2;
  repeated Span spans = 3;
}
```